From f3ec35b461a06f1278383d2347dfbc611b6a91a6 Mon Sep 17 00:00:00 2001
From: Rajasekar Raja <rajasekarr@nvidia.com>
Date: Fri, 17 May 2024 12:36:31 -0700
Subject: [PATCH] bgpd: backpressure - Fix to withdraw evpn type-5 routes
 immediately

As part of backpressure changes, there is a bug where immediate withdraw
is to be sent for evpn imported type-5 prefix to clear the nh neigh and
RMAC entry.

Fixing this by sending withdraw immediately to keep it inline with the
code today

Ticket: #3905571

Signed-off-by: Donald Sharp <sharpd@nvidia.com>
Signed-off-by: Rajasekar Raja <rajasekarr@nvidia.com>

diff --git a/bgpd/bgp_route.c b/bgpd/bgp_route.c
index 679abba463..e28069767f 100644
--- a/bgpd/bgp_route.c
+++ b/bgpd/bgp_route.c
@@ -3312,8 +3312,7 @@ static void bgp_process_main_one(struct bgp *bgp, struct bgp_dest *dest,
 			 */
 			if (old_select &&
 			    is_route_parent_evpn(old_select))
-				bgp_zebra_route_install(dest, old_select, bgp,
-							false, NULL, false);
+				bgp_zebra_withdraw_actual(dest, old_select, bgp);
 
 			bgp_zebra_route_install(dest, new_select, bgp, true,
 						NULL, false);
diff --git a/bgpd/bgp_zebra.c b/bgpd/bgp_zebra.c
index 524551b1e0..5d5525156b 100644
--- a/bgpd/bgp_zebra.c
+++ b/bgpd/bgp_zebra.c
@@ -1889,11 +1889,9 @@ static void bgp_zebra_buffer_write_ready(void)
  *                                     save new pi, mark as going to be
  *                                     withdrawan, remove install flag
  *
- *    Withdrawal       Install         Special case, send withdrawal immediately
- *                                     Leave dest on list, release old pi,
+ *    Withdrawal       Install         Leave dest on list, release old pi,
  *                                     save new pi, mark as going to be
- *                                     installed.  <see note about evpn
- *                                     in bgp_route.c in bgp_process_main_one>
+ *                                     installed.
  *    Withdrawal       Withdrawal      Leave dest on list, release old pi,
  *                                     save new pi, mark as going to be
  *                                     withdrawn.
@@ -1949,9 +1947,6 @@ void bgp_zebra_route_install(struct bgp_dest *dest, struct bgp_path_info *info,
 		dest->za_bgp_pi = info;
 	} else if (CHECK_FLAG(dest->flags, BGP_NODE_SCHEDULE_FOR_DELETE)) {
 		assert(dest->za_bgp_pi);
-		if (install & !is_evpn)
-			bgp_zebra_withdraw_actual(dest, dest->za_bgp_pi, bgp);
-
 		bgp_path_info_unlock(dest->za_bgp_pi);
 		bgp_path_info_lock(info);
 		dest->za_bgp_pi = info;
-- 
2.43.2

